woocommerce/google-listings-and-ads
# Attribute Mapping - Mapping function

### Changes proposed in this Pull Request:

Closes part of #1648 .

This PR implements the mapping function for Product Syncing.

a. For individual product sync 
b. For batch syncing

It implements the next mappings:

1.  Fixed values
2.  Enum values
3.  Taxonomies
4.  Product fields
5.  Custom created attributes 
6. Yoast Seo GTIN and MPN 

Notice Attributes set in GLA Attributes at product level have priority over rules. So on esting be sure the value is not overrided by some attributes already set in your product

### Detailed test instructions:


- Empty al your rules in `wp_gla_attribute_mapping_rules `
- check out this PR and Go to Attribute Mapping
- For testing the sync you can use for now connection test page
- You can use debugger to see the final data or just check in your Merchant Center. (notice it takes a few minutes to update)

**Testing Category Conditions**

1. Add one rule with Category condition ALL. Check that the rule applies to all the products
2. Add a category in some products. Add one rule with Category condition ONLY and that category . Check that the rule only applies to the products with that category.
3. Add one rule with Category condition EXCEPT and that category . Check that the rule only applies to the products that are not in that category.

**Testing product field rules:**

1. Create type ALL rules using this fields: Allow backorders, Product title, SKU, Stock Qty, Stock Status, Tax class, Variation title, Weight (raw value, no units), Weight (with units). See that the values are properly loaded in the attribute of your choice.

**Testing taxonomy rules:**

1. Create type ALL rules using Color taxomomy, Size taxonomy and WooCommerce Brands (you should install woocommerce brands)
2. Create a new attribute and see that it works as well.

**Testing YOAST rules:**

 1. Follow https://github.com/woocommerce/google-listings-and-ads/pull/428 to simulate you have WC YOAST installed
 2. Test GTIN Yoast and MPN Yoast for MPN and GTIN

**Testing fixed values**

1. Assign a fixed value to some rule and see it works. 

**Testing Custom Hook attributes** 

1. Create an attribute at product level
2. Load that attribute via filter hook `woocommerce_gla_attribute_mapping_sources_custom_attributes`. Example

```
add_filter(
		'woocommerce_gla_attribute_mapping_sources_custom_attributes',
		function( $values ) {
			return ['total_sales'];
		}
	);
```

3. Test that the attribute appears in the selector and works

**Testing GLA Attributes priority** 

1. Add some attribute in GLA Attributes Tab for a product.
2. Create some rule that applies that attribute. 
3. See how at the end GLA Attribute takes place.
